### ISSUE TITLE
Azure Storage Explorer ACL

### ISSUE THREAD

**Issue by kapmib** [microsoft/AzureStorageExplorer#6354]
### Preflight Checklist

- [X] I have installed the [latest version of Storage Explorer](https://github.com/Microsoft/AzureStorageExplorer/releases/latest).
- [X] I have checked existing resources, including the [troubleshooting guide](https://docs.microsoft.com/azure/storage/common/storage-explorer-troubleshooting) and the [release notes](https://github.com/microsoft/AzureStorageExplorer/releases).
- [X] I have searched for [similar issues](https://github.com/Microsoft/AzureStorageExplorer/issues).

### Storage Explorer Version

1.27.0

### Architecture

i86

### Storage Explorer Build Number

20221201.1

### Platform

All

### OS Version

Windows 10

### Bug Description

Generated SAS token for ADLS Gen2 Storage account to B2B user with LIST permission.
Added B2B user to ACL on a second level Directory with READ, WRITE, EXECUTE permission and set "Propagate Access Control Lists" and checked that the ACL has been propagated to all children. 
The user is able to view directories and files, but get "Failed to start transfer: Insufficient credentials" error when trying to download/open file.
When generating a SAS token with LIST and READ permission, the user is able to download/open the file.

### Steps to Reproduce

Generate SAS token from Azure Storage Account with LIST permission
Open Azure Data Explorer and connect to Azure ADLS Gen2 container with SAS token.
Try to download/open file from one of the children folders/directories

### Actual Experience

Getting error: "Failed to start transfer: Insufficient credentials" 

### Expected Experience

I want the B2B user to access the blob container through a SAS token but at the same time manage which files can be downloaded and uploaded through ACL.

**Comment by MRayermannMSFT** [comment]
> The user is able to view directories and files, but get "Failed to start transfer: Insufficient credentials" error when trying to download/open file.

Can you please double check the user's permissions on the specified file?

**Comment by kapmib** [comment]
> 

The user has Read, Write and Execute on ACL on the specified file as well. It seems like the ACL settings isn't doing anything as it can view all folders and files through the List permission on it's own from the SAS token.

Shouldn't my approach work? I.e. to forward a SAS token to the user with LIST permission for the ADLS Gen2 container and then giving Write, Read and Exucute permission through ACL on a second level folder and the same as well for the nested folder and file.

**Comment by kapmib** [comment]
> > The user is able to view directories and files, but get "Failed to start transfer: Insufficient credentials" error when trying to download/open file.
> 
> Can you please double check the user's permissions on the specified file?

The user has Read, Write and Execute on ACL on the specified file as well. It seems like the ACL settings isn't doing anything as it can view all folders and files through the List permission on it's own from the SAS token.

Shouldn't my approach work? I.e. to forward a SAS token to the user with LIST permission for the ADLS Gen2 container and then giving Write, Read and Exucute permission through ACL on a second level folder and the same as well for the nested folder and file.

**Comment by MRayermannMSFT** [comment]
> Shouldn't my approach work? I.e. to forward a SAS token to the user with LIST permission for the ADLS Gen2 container and then giving Write, Read and Exucute permission through ACL on a second level folder and the same as well for the nested folder and file.

No this won't work. If the user has added the connection to Storage Explorer with a SAS URL then requests sent to the server are authenticated with the SAS token. There's no way for the server to know that the user is the user.

It sounds like your best option is to have the user add their ADLS Gen2 container connection to Storage Explorer via Azure AD. 
![image](https://user-images.githubusercontent.com/21112697/207129566-3d834f7d-6f5d-42b4-8662-38cad5f4958f.png)

- This will mean the user needs to first sign into Storage Explorer, then go back to the connect dialog and do the connection process.
- After picking their account and tenant, the user will need to provide a URL. Since you are using ADLS Gen2, they can simply use the URL of the container+the path of the directory containing the file you want them to download
- For ACLs, the user will at a minimum need:
   - Execute on every directory above the directory containing the file you want them to download
   - Execute and read on the directory containing the file you want them to download
   - Read on the file you want them to download

**Comment by kapmib** [comment]
> > Shouldn't my approach work? I.e. to forward a SAS token to the user with LIST permission for the ADLS Gen2 container and then giving Write, Read and Exucute permission through ACL on a second level folder and the same as well for the nested folder and file.
> 
> No this won't work. If the user has added the connection to Storage Explorer with a SAS URL then requests sent to the server are authenticated with the SAS token. There's no way for the server to know that the user is the user.
> 
> It sounds like your best option is to have the user add their ADLS Gen2 container connection to Storage Explorer via Azure AD. ![image](https://user-images.githubusercontent.com/21112697/207129566-3d834f7d-6f5d-42b4-8662-38cad5f4958f.png)
> 
> * This will mean the user needs to first sign into Storage Explorer, then go back to the connect dialog and do the connection process.
> * After picking their account and tenant, the user will need to provide a URL. Since you are using ADLS Gen2, they can simply use the URL of the container+the path of the directory containing the file you want them to download
> * For ACLs, the user will at a minimum need:
>   
>   * Execute on every directory above the directory containing the file you want them to download
>   * Execute and read on the directory containing the file you want them to download
>   * Read on the file you want them to download

Thank you for your reply. I wanted to test your approach using my own user (Contributor) to access the Gen 2 blob container through _Sign in using Azure Active Directory (Azure AD)_

My thesis is that I should be able to access the container with my own user before starting testing on the B2B user as I already have access to everything. 
But when I enter _Blob container or directory URL_ using your guide I am getting the following error - please note that I got the URL from properties on my ADLS container (see attachment 1):
 "Server failed to authenticate the request. Please refer to the information in the www-authenticate header. RequestId:30931fa3-801e-000a-4245-0f3981000000 "
 
Where am I going wrong? Am I using the wrong _Blob container or directory URL_ when trying to _Sign in using Azure Active Directory (Azure AD)_ ?
![1 - landingzone ADLS Gen 2 blob URL](https://user-images.githubusercontent.com/120135894/207463661-632642e3-92c2-420e-8e4c-9218016aa79c.png)
![2 - landingzone ADLS Gen 2 AAD connection info](https://user-images.githubusercontent.com/120135894/207463664-3f90a8b9-1cf0-4730-8808-c068ea032b88.png)
![3 - landingzone after Azure AD sign in](https://user-images.githubusercontent.com/120135894/207463667-20bf2e84-ff3b-4771-b69b-2b86bfb1a4ad.png)

**Comment by JasonYeMSFT** [comment]
@kapmib Please make sure in the "Select Account & Tenant" step, you are selecting an account-tenant pair that grants you access to the blob container. If you believe that's correct, please follow the steps to raise the debug level to DEBUG and [Gather Storage Explorer logs](https://learn.microsoft.com/en-us/azure/storage/common/storage-explorer-troubleshooting?tabs=Windows#storage-explorer-logs). The logs will redact all your credentials, if you are concerned with exposing your storage account names, you can email it to sehelp at microsoft dot com.

**Comment by MRayermannMSFT** [comment]
Closing due to lack of response. If you are still running into this problem we recommend opening a support ticket from the Azure portal. You can also open a new issue.
